osu.Game.Rulesets.Catch/UI/CatcherTrailDisplay.cs

```diff
@@ -17,7 +17,14 @@ namespace osu.Game.Rulesets.Catch.UI
     /// </summary>
     public class CatcherTrailDisplay : CompositeDrawable
     {
-        private readonly Catcher catcher;
+        /// <summary>
+        /// The most recent dash trail added in this container.
+        /// Only alive (not faded out) trails are considered.
+        /// </summary>
+        [CanBeNull]
+        public CatcherTrail LastDashTrail => dashTrails.Concat(hyperDashTrails)
+                                                       .OrderByDescending(trail => trail.LifetimeStart)
+                                                       .FirstOrDefault();
```

Comment:
Looks quite expensive for something to be read from `Catcher` every frame, especially so that doing `dashTrails.Add(LastDashTrail = trail)` (and the exact same thing for hyperdash trails) lower down looks like a viable O(1) alternative.

Reply:
Using a variable to store `LastDashTrail` doesn't work when a replay is rewound.

Reply:
At very least, if this is to be run every frame it should not have any LINQ (to avoid the allocation overhead).